mantidproject/mantid
# Allow `SofQW` to rebin in energy and speed it up

The three `SofQW` algorithms have been modified to allow them to rebin an input inelastic dataset in (spectrum-number, `DeltaE`) to (`MomentumTransfer`, `DeltaE`) with arbitrary `DeltaE` and `MomentumTransfer` bins. Previously, the `DeltaE` axis was left unchanged by the rebinning.

In addition, `FractionalRebinning` has been modified to add two extra routines for calculating the polygon overlaps which `SofQWPolygon` and `SofQWNormalisedPolygon` needs. The new routines are for cases when the input grid is axis-aligned with the output grid (e.g. for `Rebin2D`), or for cases where the vertical (`y`-) axis is aligned in the two grids (the case for `SofQW*` on account of the `DeltaE` axis). The rectangular (both axes aligned) routine is much faster (>2-3 times) than the original routine (which works for arbitrary shaped polygons). The trapezoidal (one axis aligned) routine is faster, but only by ~50%, depending on the size of the data. The remaining computing time is due to overheads in setting up vectors and allocating data (as deduced from valgrind / `gprof` / `sprof` output) and would be difficult to elliminate.

**To test:**

Code review and check automated tests run ok. A new test for the energy rebinning of `SofQW*` has been added.

Run through a direct-inelastic or reflectometry workflow (e.g. from instructions to test PR #20953) 

Fixes #21500

**Release Notes** 


Release notes added.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
